README.md
# numer.js

Numer.js is an open-source JavaScript library for formatting and manipulating numbers.

[![NPM](https://img.shields.io/npm/v/numer.js.svg)](https://www.npmjs.com/package/numer.js) ![npm bundle size](https://img.shields.io/bundlephobia/min/numer.js) [![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

## 🎁 Features

* Numer() constructor
  * format() instance method

## 🔧 Install

numer.js is available on npm. It can be installed with the following command:

```
npm install numer.js --save
```

numer.js is available on yarn as well. It can be installed with the following command:

```
yarn add numer.js --save
```

If you don't want to use npm or yarn, [numer.min.js](https://unpkg.com/numer.js@0.1.0/numer.min.js) can be downloaded to your project source.

## 📖 Documentation

### Constructor

* Numer(options)
  * Creates a new Numer object.

```js
new Numer(options)
```

#### Parameters

* options
  * An object with some or all of the following properties.

#### style

The formatting style to use.

* comma for comma formatting.
* abbreviation for abbreviation formatting.
* ordinal for ordinal formatting.

### Instance methods

* Numer.prototype.format(number)
  * Getter function that formats a number according to the formatting options of this Numer object.

```js
format(number)
```

#### Parameters

* number
  * A Number or BigInt to format.

## 💡 Usage

### 🎀 format()

```js
// Node.js
const Numer = require("numer.js");

console.log(new Numer({ style: 'comma' }).format(1000000));
// expected output: "1,000,000"

console.log(new Numer({ style: 'abbreviation' }).format(9812730));
// expected output: "9.8M"

console.log(new Numer({ style: 'ordinal' }).format(1));
// expected output: "1st"
```

## 📜 Changelog

Latest version 0.1.0 (2021-09-07):

  * Numer() constructor
    * format() instance method

Details changes for each release are documented in the [CHANGELOG.md](https://github.com/Bunlong/numer.js/blob/master/CHANGELOG.md).

## ❗ Issues

If you think any of the `numer.js` can be improved, please do open a PR with any updates and submit any issues. Also, I will continue to improve this, so you might want to watch/star this repository to revisit.

## 💪 Contribution

We'd love to have your helping hand on contributions to `numer.js` by forking and sending a pull request!

Your contributions are heartily ♡ welcome, recognized and appreciated. (✿◠‿◠)

How to contribute:

- Open pull request with improvements
- Discuss ideas in issues
- Spread the word
- Reach out with any feedback

## 🏆 Contributors

<table>
  <tr>
    <td align="center">
      <a href="https://github.com/Bunlong">
        <img src="https://avatars0.githubusercontent.com/u/1308397?s=400&u=945dc6b97571e2b98b659d34b1c81ae2514046bf&v=4" width="100" alt="Bunlong" />
        <br />
        <sub>
          <b>Bunlong</b>
        </sub>
      </a>
    </td>
  </tr>
</table>

## ⚖️ License

The MIT License [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)


CHANGELOG.md
# 0.1.0 (2021-09-07)

## Features

* Numer() constructor
  *format() instance method

## Credit

* [Bunlong](https://github.com/Bunlong)


.eslintrc.js
module.exports = {
  "parserOptions": {
    "ecmaVersion": 5
  },
  "env": {
    "browser": true,
    "worker": true,
    "node": true
  },
  "extends": "eslint:recommended",
  "rules": {
    "accessor-pairs": "error",
    "array-bracket-newline": ["error", "consistent"],
    "array-bracket-spacing": [
      "error",
      "never"
    ],
    "array-callback-return": "error",
    "array-element-newline": "off",
    "arrow-body-style": "error",
    "arrow-parens": "error",
    "arrow-spacing": "error",
    "block-scoped-var": "error",
    "block-spacing": "error",
    "brace-style": "off",
    "callback-return": "error",
    "camelcase": ["error", {"properties": "never"}],
    "capitalized-comments": "off",
    "class-methods-use-this": "error",
    "comma-dangle": "off",
    "comma-spacing": "off",
    "comma-style": [
      "error",
      "last"
    ],
    "complexity": "off",
    "computed-property-spacing": [
      "error",
      "never"
    ],
    "consistent-return": "off",
    "consistent-this": "off",
    "curly": "off",
    "default-case": "error",
    "dot-location": "error",
    "dot-notation": "error",
    "eol-last": "error",
    "eqeqeq": "error",
    "for-direction": "error",
    "func-call-spacing": "error",
    "func-name-matching": "error",
    "func-names": [
      "error",
      "never"
    ],
    "func-style": "off",
    "function-paren-newline": "off",
    "generator-star-spacing": "error",
    "getter-return": "error",
    "global-require": "off",
    "guard-for-in": "off",
    "handle-callback-err": "error",
    "id-blacklist": "error",
    "id-length": "off",
    "id-match": "error",
    "implicit-arrow-linebreak": "error",
    "indent": "off",
    "indent-legacy": "off",
    "init-declarations": "off",
    "jsx-quotes": "error",
    "key-spacing": "error",
    "keyword-spacing": "off",
    "line-comment-position": "off",
    "linebreak-style": [
      "error",
      "unix"
    ],
    "lines-around-comment": "off",
    "lines-around-directive": "off",
    "lines-between-class-members": "error",
    "max-depth": "off",
    "max-len": "off",
    "max-lines": "off",
    "max-nested-callbacks": "error",
    "max-params": "off",
    "max-statements": "off",
    "max-statements-per-line": "off",
    "multiline-comment-style": "off",
    "multiline-ternary": [
      "error",
      "always-multiline"
    ],
    "new-parens": "error",
    "newline-after-var": "off",
    "newline-before-return": "off",
    "newline-per-chained-call": "off",
    "no-alert": "error",
    "no-array-constructor": "error",
    "no-await-in-loop": "error",
    "no-bitwise": "error",
    "no-buffer-constructor": "error",
    "no-caller": "error",
    "no-catch-shadow": "off",
    "no-cond-assign": [
      "error",
      "except-parens"
    ],
    "no-confusing-arrow": "error",
    "no-console": "off",
    "no-continue": "off",
    "no-div-regex": "error",
    "no-duplicate-imports": "error",
    "no-else-return": "off",
    "no-empty": ["error", {"allowEmptyCatch": true}],
    "no-empty-function": "off",
    "no-eq-null": "error",
    "no-eval": "error",
    "no-extend-native": "error",
    "no-extra-bind": "error",
    "no-extra-label": "error",
    "no-extra-parens": "off",
    "no-floating-decimal": "error",
    "no-implicit-globals": "error",
    "no-implied-eval": "error",
    "no-inline-comments": "off",
    "no-inner-declarations": [
      "error",
      "functions"
    ],
    "no-invalid-this": "off",
    "no-iterator": "error",
    "no-label-var": "error",
    "no-labels": "error",
    "no-lone-blocks": "error",
    "no-lonely-if": "error",
    "no-loop-func": "error",
    "no-magic-numbers": "off",
    "no-mixed-operators": "off",
    "no-mixed-requires": "error",
    "no-multi-assign": "error",
    "no-multi-spaces": "off",
    "no-multi-str": "error",
    "no-multiple-empty-lines": "off",
    "no-native-reassign": "error",
    "no-negated-condition": "off",
    "no-negated-in-lhs": "error",
    "no-nested-ternary": "off",
    "no-new": "error",
    "no-new-func": "error",
    "no-new-object": "error",
    "no-new-require": "error",
    "no-new-wrappers": "error",
    "no-octal-escape": "error",
    "no-param-reassign": "off",
    "no-path-concat": "off",
    "no-plusplus": "off",
    "no-process-env": "error",
    "no-process-exit": "error",
    "no-proto": "error",
    "no-prototype-builtins": "error",
    "no-restricted-globals": "error",
    "no-restricted-imports": "error",
    "no-restricted-modules": "error",
    "no-restricted-properties": "error",
    "no-restricted-syntax": "error",
    "no-return-assign": "error",
    "no-return-await": "error",
    "no-script-url": "error",
    "no-self-compare": "error",
    "no-sequences": "error",
    "no-shadow": "off",
    "no-shadow-restricted-names": "error",
    "no-spaced-func": "error",
    "no-sync": ["error", {"allowAtRootLevel": true}],
    "no-tabs": "off",
    "no-template-curly-in-string": "error",
    "no-ternary": "off",
    "no-throw-literal": "error",
    "no-trailing-spaces": "error",
    "no-undef-init": "error",
    "no-undefined": "off",
    "no-underscore-dangle": "off",
    "no-unmodified-loop-condition": "off",
    "no-unneeded-ternary": "error",
    "no-unused-expressions": "off",
    "no-unused-vars": ["error", {"args": "none"}],
    "no-use-before-define": "off",
    "no-useless-call": "error",
    "no-useless-computed-key": "error",
    "no-useless-concat": "error",
    "no-useless-constructor": "error",
    "no-useless-rename": "error",
    "no-useless-return": "off",
    "no-var": "off",
    "no-void": "error",
    "no-warning-comments": "error",
    "no-whitespace-before-property": "error",
    "no-with": "error",
    "nonblock-statement-body-position": [
      "error",
      "any"
    ],
    "object-curly-newline": ["error", {"consistent": true}],
    "object-curly-spacing": "off",
    "object-shorthand": "off",
    "one-var": "off",
    "one-var-declaration-per-line": "off",
    "operator-assignment": [
      "error",
      "always"
    ],
    "operator-linebreak": "off",
    "padded-blocks": "off",
    "padding-line-between-statements": "error",
    "prefer-arrow-callback": "off",
    "prefer-const": "error",
    "prefer-destructuring": "off",
    "prefer-numeric-literals": "error",
    "prefer-promise-reject-errors": "error",
    "prefer-reflect": "off",
    "prefer-rest-params": "off",
    "prefer-spread": "error",
    "prefer-template": "off",
    "quote-props": "off",
    "quotes": "off",
    "radix": [
      "error",
      "as-needed"
    ],
    "require-await": "error",
    "require-jsdoc": "off",
    "rest-spread-spacing": "error",
    "semi": "error",
    "semi-spacing": "error",
    "semi-style": [
      "error",
      "last"
    ],
    "sort-imports": "error",
    "sort-keys": "off",
    "sort-vars": "off",
    "space-before-blocks": "error",
    "space-before-function-paren": [
      "error",
      "never"
    ],
    "space-in-parens": [
      "error",
      "never"
    ],
    "space-infix-ops": "error",
    "space-unary-ops": [
      "error",
      {
        "nonwords": false,
        "words": false
      }
    ],
    "spaced-comment": "off",
    "strict": "off",
    "switch-colon-spacing": "error",
    "symbol-description": "error",
    "template-curly-spacing": "error",
    "template-tag-spacing": "error",
    "unicode-bom": [
      "error",
      "never"
    ],
    "valid-jsdoc": "off",
    "vars-on-top": "off",
    "wrap-iife": "off",
    "wrap-regex": "off",
    "yield-star-spacing": "error",
    "yoda": "off"
  }
};


numeric-format.js
(function(root, factory) {
  /* globals define */
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define([], factory);
  } else if (typeof module === 'object' && typeof exports !== 'undefined') {
    // Node. Does not work with strict CommonJS, but
    // only CommonJS-like environments that support module.exports,
    // like Node.
    module.exports = factory();
  } else {
    // Browser globals (root is window)
    root.NumericFormat = factory();
  }
})(this, function() { 'use strict';
  var global = (function() {
    // alternative method, similar to `Function('return this')()`
    // but without using `eval` (which is disabled when
    // using Content Security Policy).

    if (typeof self !== 'undefined') { return self; }
    if (typeof window !== 'undefined') { return window; }
    if (typeof global !== 'undefined') { return global; }

    // When running tests none of the above have been defined
    return {};
  })();

  var NumericFormat = {};

  NumericFormat.numberWithCommas = numberWithCommas;
  NumericFormat.compactNumber = compactNumber;
  NumericFormat.ordinalSuffix = ordinalSuffix;
  NumericFormat.getRandomInt = getRandomInt;

  function numberWithCommas(_number) {
    return _number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  }

  function compactNumber(_number) {
    var suffixes = ['', 'k', 'm', 'b', 't'];
    var suffixNum = Math.floor(('' + _number).length / 3);
    var shortValue = parseFloat(
      (suffixNum !== 0 ? _number / Math.pow(1000, suffixNum) : _number).toPrecision(2)
    );
    if (shortValue % 1 !== 0) {
      shortValue = shortValue.toFixed(1);
    }
    return shortValue + suffixes[suffixNum];
  }

  function ordinalSuffix(_number) {
    var j = _number % 10;
    var  k = _number % 100;
    if (j === 1 && k !== 11) {
      return _number + 'st';
    }
    if (j === 2 && k !== 12) {
      return _number + 'nd';
    }
    if (j === 3 && k !== 13) {
      return _number + 'rd';
    }
    return _number + 'th';
  }

  function getRandomInt(_min, _max) {
    var min = Math.ceil(_min);
    var max = Math.floor(_max);
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  return NumericFormat;
});


numer.js
/* @license
numer.js
v0.1.0
https://github.com/bunlong/numer.js
License: MIT
*/

(function(root, factory) {
  /* globals define */
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define([], factory);
  } else if (typeof module === 'object' && typeof exports !== 'undefined') {
    // Node. Does not work with strict CommonJS, but
    // only CommonJS-like environments that support module.exports,
    // like Node.
    module.exports = factory();
  } else {
    // Browser globals (root is window)
    root.Numer = factory();
  }
})(this, function() {
  'use strict';

  function Numer(input) {
    this._input = input;
  }

  Numer.prototype = {
    format: function(number) {
      var _input = this._input;
      var _number = number;
      switch (_input.style) {
        case 'comma':
          return addCommas(_number);
        case 'abbreviation':
          return abbreviate(_number);
        case 'ordinal':
          return convertToOrdinal(_number);
        default:
          throw new Error(
            'The formatting style to use: comma, abbreviation and ordinal.'
          );
      }
    },
  };

  function addCommas(_number) {
    return _number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  }

  function abbreviate(_value) {
    var _newValue = _value;
    if (_value >= 1000) {
      var _suffixes = ['', 'K', 'M', 'B', 'T'];
      var _suffixNum = Math.floor(('' + _value).length / 3);
      var _shortValue = '';
      for (var precision = 2; precision >= 1; precision--) {
        _shortValue = parseFloat(
          (_suffixNum !== 0
            ? _value / Math.pow(1000, _suffixNum)
            : _value
          ).toPrecision(precision)
        );
        var dotLessShortValue = (_shortValue + '').replace(
          /[^a-zA-Z 0-9]+/g,
          ''
        );
        if (dotLessShortValue.length <= 2) {
          break;
        }
      }
      if (_shortValue % 1 !== 0) _shortValue = _shortValue.toFixed(1);
      _newValue = _shortValue + _suffixes[_suffixNum];
    }
    return _newValue;
  }

  function convertToOrdinal(_number) {
    var _j = _number % 10;
    var _k = _number % 100;
    if (_j === 1 && _k !== 11) {
      return _number + 'st';
    }
    if (_j === 2 && _k !== 12) {
      return _number + 'nd';
    }
    if (_j === 3 && _k !== 13) {
      return _number + 'rd';
    }
    return _number + 'th';
  }

  return Numer;
});
